### ISSUE TITLE
Dafny claims (b * a) % a != 0 and provides an incorrect counterexample

### ISSUE THREAD

**Issue by johnterickson** [dafny-lang/dafny#4246]
### Dafny version

4.1.0

### Code to produce this issue

```dafny
lemma BasicMath(b: nat, a: nat)
    requires a != 0
    ensures b*a/a == b
{
    var p := b * a;
    assert p % a == 0; // assert fails. 
    assert p / a == b;
}
```

### Command to run and resulting output

```code
PS C:\src\DafnyBook> & "C:\Program Files\dotnet\dotnet.EXE" c:\Users\jerick\.vscode\extensions\dafny-lang.ide-vscode-3.1.1\out\resources\4.1.0\github\dafny\Dafny.dll run "c:\src\DafnyBook\repro.dfy"
c:\src\DafnyBook\repro.dfy(6,17): Error: assertion might not hold

Dafny program verifier finished with 1 verified, 1 error

// counterexample returned (see screenshot) is b == 147, a == 250, p == 36750.  But p % a == 36750 % 250 == 0.
```

### What happened?

![image](https://github.com/dafny-lang/dafny/assets/10488772/f35fe56d-c108-4eb6-8f1b-3e65f2fa07a0)

### What type of operating system are you experiencing the problem on?

Windows

**Comment by johnterickson** [comment]
An even simpler example: `p := b * a` but `998001` != `1` * `999`

```
lemma BasicMath(b: nat, a: nat)
    requires 0 < a < 1000
    requires 0 < b < 1000
{
    var p := b * a;
    var quotient := p / a;
    assert quotient * a == p;
}
```

![image](https://github.com/dafny-lang/dafny/assets/10488772/2d7530ac-c7af-49e7-ab21-65d1c9abea15)

**Comment by MikaelMayer** [comment]
Thank you for reporting. we don't know how to advertise it, but Dafny is a verification language, so when Dafny says it's verified, you can trust it. But counter-examples are just what we retrieve from a solver which was not able to verify the code. They might or might not be real counter-examples, it's just that the solver wasn't able to disprove the validify of that counter example, because of the encoding.
So I apologize for the confusion. Counter-examples are tricky as it's not always possible to run them, if it includes ghost state for example.

**Comment by johnterickson** [comment]
Makes sense - thanks for the explanation!  I expected that counter-examples are not guaranteed to disprove and were more of "the solver got stuck/confused in this situation" but I am surprised that it's not even internally consistent with itself.

That said, I also don't have a great answer for how to advertise this functionality. "Counter-example" does seem to be a much stronger claimer than is currently offered. Perhaps "solver breadcrumbs" but that is a pretty culturally specific idiom. https://en.wiktionary.org/wiki/breadcrumb

Maybe "solver dead-ends"?

Anyway, feel free to close 🙂

**Comment by johnterickson** [comment]
PS thank you for the prompt response!
